<a href="https://colab.research.google.com/github/lakesidetech/machine_learning/blob/main/titanic_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***This program predicts if a passenger will survive on the titanic***

# ***import liabries***

In [6]:
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
import numpy as np
from pandas import read_csv


In [5]:
#load the data into dataframe

train_titanic =pd.read_csv('/content/train.csv')
test_titanic =pd.read_csv('/content/test.csv')

# Exploratory Data Analysis (EDA)

In [7]:
#Check data your column's datatype
#check for empty or missing value
train_titanic.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


The train_titanic data-set has 891 examples and 11 features + the target variable (survived). 2 of the features are floats, 5 are integers and 5 are objects.


In [8]:
# data shape
# get number of rows and columns in the dataset
train_titanic.shape
#test_titanic.shape


(891, 12)

In [9]:
# data shape
## get number of rows and columns in the dataset

test_titanic.shape


(418, 11)

In [17]:
train_titanic.tail

<bound method NDFrame.tail of      PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0              1         0       3  ...   7.2500   NaN         S
1              2         1       1  ...  71.2833   C85         C
2              3         1       3  ...   7.9250   NaN         S
3              4         1       1  ...  53.1000  C123         S
4              5         0       3  ...   8.0500   NaN         S
..           ...       ...     ...  ...      ...   ...       ...
886          887         0       2  ...  13.0000   NaN         S
887          888         1       1  ...  30.0000   B42         S
888          889         0       3  ...  23.4500   NaN         S
889          890         1       1  ...  30.0000  C148         C
890          891         0       3  ...   7.7500   NaN         Q

[891 rows x 12 columns]>

In [18]:
# Looking at the PassengerId   column
print (train_titanic['PassengerId'])
print (train_titanic['PassengerId'].isnull())


0        1
1        2
2        3
3        4
4        5
      ... 
886    887
887    888
888    889
889    890
890    891
Name: PassengerId, Length: 891, dtype: int64
0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: PassengerId, Length: 891, dtype: bool


In [12]:
#check and drop duplicate rows in your dataset
duplicated_rows=train_titanic[train_titanic.duplicated()]
print('dupicates are:',duplicated_rows)

dupicates are: Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []


In [23]:
total = train_titanic.isnull().sum().sort_values(ascending=False)
# sumof_train_dataset/count_train_dataset*100
percent_1 = train_titanic.isnull().sum()/train_titanic.isnull().count()*100 
# round the percent to 1 decimal place then sort result from highest to lowest (descending)
percent_2 = (round(percent_1, 1)).sort_values(ascending=False)
missing_titanic_data = pd.concat([total, percent_2], axis=1, keys=['Column Total', '%'])
missing_titanic_data


Column Total     %
Cabin                 687  77.1
Age                   177  19.9
Embarked                2   0.2
Fare                    0   0.0
Ticket                  0   0.0
Parch                   0   0.0
SibSp                   0   0.0
Sex                     0   0.0
Name                    0   0.0
Pclass                  0   0.0
Survived                0   0.0
PassengerId             0   0.0

missing values embarked 2 @ 0.2%, Age 177 @ 19.9% and Cabin 687 @77.1%

clean titanic train_dataset

In [12]:
#Check which columns contain empty values (NaN, NAN, na). Looks like columns age, embarked, deck, and embarked_town are missing some values.
#Count the empty (NaN, NAN, na) values in each column train_titanic.isna().sum()
train_titanic.isna().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [9]:
#get some statistics on the dataset\
train_titanic.describe()


PassengerId    Survived      Pclass  ...       SibSp       Parch        Fare
count   891.000000  891.000000  891.000000  ...  891.000000  891.000000  891.000000
mean    446.000000    0.383838    2.308642  ...    0.523008    0.381594   32.204208
std     257.353842    0.486592    0.836071  ...    1.102743    0.806057   49.693429
min       1.000000    0.000000    1.000000  ...    0.000000    0.000000    0.000000
25%     223.500000    0.000000    2.000000  ...    0.000000    0.000000    7.910400
50%     446.000000    0.000000    3.000000  ...    0.000000    0.000000   14.454200
75%     668.500000    1.000000    3.000000  ...    1.000000    0.000000   31.000000
max     891.000000    1.000000    3.000000  ...    8.000000    6.000000  512.329200

[8 rows x 7 columns]

# Build the machine model and compare the results. 


Stochastic Gradient Descent (SGD):

Random Forest


Logistic Regression


K Nearest Neighbor


Gaussian Naive Bayes


Support Vector Machine


Perceptron


Decision Tree


Conclude and tell us what would be the best Model in this respect to use?